### DISCUSSION TITLE
DataTable conditional formatting?

### DISCUSSION THREAD

**Discussion by TomJGooding** [Textualize/textual#2101] [Q&A]
How might you approach applying some form of 'conditional formatting' to a DataTable?

Below is a contrived example of a table with student exam results data. Let's say that results above 90% should be formatted in green, below 30% in red, and any results in between in orange.

I'm still not sure I fully understand just how to construct a DataTable -  hopefully the docs for this might be expanded at some point?

Many thanks in advance for any advice!

```python
from textual.app import App, ComposeResult
from textual.widgets import DataTable

ROWS = [
    ("Student", "Result"),
    ("John Smith", 91),
    ("Steve McMara", 56),
    ("Tim Bow", 24),
    ("Frank Bush", 27),
    ("Walter Obama", 99),
    ("Rachel Clinton", 23),
    ("Monica Diesel", 55),
    ("Ted Wilder", 10),
]


class TableApp(App):
    def compose(self) -> ComposeResult:
        yield DataTable()

    def on_mount(self) -> None:
        table = self.query_one(DataTable)
        rows = iter(ROWS)
        table.add_columns(*next(rows))
        table.add_rows(rows)


app = TableApp()
if __name__ == "__main__":
    app.run()
```

**Comment by TomJGooding** [comment]
After spending some time reading other discussions and the docs in more detail, I've realised that you can use Rich console markup to apply formatting to `DataTable` cells.

Perhaps there is still a way of formatting table cells with CSS that I'm missing?

My updated code below is very inefficient as it creates then immediately updates each cell, but I'm still trying to get to grips with this!

```python
    ...
    
    def on_mount(self) -> None:
        table = self.query_one(DataTable)
        rows = iter(ROWS)
        table.add_columns(*next(rows))
        table.add_rows(rows)

        self.format_results()

    def format_results(self) -> None:
        table = self.query_one(DataTable)
        result_cells = table.get_column_at(1)
        for idx, result in enumerate(result_cells):
            cell_coordinate = Coordinate(idx, 1)
            if int(result) >= 90:
                table.update_cell_at(
                    cell_coordinate,
                    f"[green]{result}[/green]",
                )
            elif int(result) <= 30:
                table.update_cell_at(
                    cell_coordinate,
                    f"[red]{result}[/red]",
                )
            else:
                table.update_cell_at(
                    cell_coordinate,
                    f"[dark_orange]{result}[/dark_orange]",
                )


```

**Reply by willmcgugan** [reply]
The CSS works at the widget level, but datatable cells aren't widgets internally. But you can still apply styles programatically.

You can style your rows when you populate them. Cell values accept any rich renderable, including [Text](https://rich.readthedocs.io/en/latest/text.html) objects.

What would you like to see added to the docs? There's not much more to it that the code you posted.

**Reply by TomJGooding** [reply]
Thanks Will for confirming. In hindsight, I think you can blame my lack of knowledge of Python iterators rather than anything to do with the docs!

**Reply by retif** [reply]
@willmcgugan is there a way to override one of the cells color for the selected/hovered row? i have a cell with circle symbol(⬤) indicator, its color is set with [yellow]⬤[/yellow] (or other color). But for selected/hovered row it become white(with all other text of the selected row), how can i disable color change?

**Reply by TomJGooding** [reply]
@retif Try changing the [`cursor_foreground_priority`](https://textual.textualize.io/widgets/data_table/#textual.widgets.DataTable(cursor_foreground_priority))

**Reply by retif** [reply]
Awesome, thank You. Looks like what i needed
